Explain that in paths generics can't be set on both the enum and the variant

estebank opened the pull request:
```
error[E0109]: type arguments are not allowed on tuple variant `TSVariant`
  --> $DIR/enum-variant-generic-args.rs:54:29
   |
LL |     Enum::<()>::TSVariant::<()>(());
   |                 ---------   ^^ type argument not allowed
   |                 |
   |                 not allowed on tuple variant `TSVariant`
   |
   = note: generic arguments are not allowed on both an enum and its variant's path segments simultaneously; they are only valid in one place or the other
help: remove the generics arguments from one of the path segments
   |
LL -     Enum::<()>::TSVariant::<()>(());
LL +     Enum::TSVariant::<()>(());
   |
LL -     Enum::<()>::TSVariant::<()>(());
LL +     Enum::<()>::TSVariant(());
   |
```

Fix #93993.

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r\? <reviewer name> (with the `\` removed)
-->
<!-- homu-ignore:end -->

estebank committed:
Explain that in paths generics can't be set on both the enum and the variant

```
error[E0109]: type arguments are not allowed on enum `Enum` and tuple variant `TSVariant`
  --> $DIR/enum-variant-generic-args.rs:54:12
   |
LL |     Enum::<()>::TSVariant::<()>(());
   |     ----   ^^   ---------   ^^ type argument not allowed
   |     |           |
   |     |           not allowed on tuple variant `TSVariant`
   |     not allowed on enum `Enum`
   |
   = note: generic arguments are not allowed on both an enum and its variant's path segments simultaneously; they are only valid in one place or the other
help: remove the generics arguments from one of the path segments
   |
LL -     Enum::<()>::TSVariant::<()>(());
LL +     Enum::<()>::TSVariant(());
   |
```

Fix #93993.

fmease commented on tests/ui/type-alias-enum-variants/enum-variant-generic-args.stderr:
```diff
@@ -279,6 +279,8 @@ LL |     Enum::<()>::TSVariant::<()>(());
    |                 |
    |                 not allowed on tuple variant `TSVariant`
+   |
+   = note: type arguments are not allowed for enum variants; specify them on the enum itself
```
This doesn't add anything? Doesn't this just regurgitate the message and labels?

Also, generic arguments *are* permitted on variants — provided none are provided to the owning enum. `Option::None::<i32>` is legal, so is `Option::<i32>::None`. However `Option::<i32>::None::<i32>` is not. I'd say this would be a helpful thing to point out (in a concise manner ofc). E.g., having a label pointing at the parent generic args (which we have here) stating sth. along the lines of "due to this" (er, you get my point).

(On mobile rn, couldn't look more carefully)

estebank replied:
Wanted to make the smallest diff that would address this. I agree that this should likely have a structured suggestion, like others in the test file, but that would have needed more code to address (I've noticed that the amount of wrangling needed for structured suggestions pushes "obviously good wanted changes" to "we might not want to land this").

estebank commented on compiler/rustc_hir_analysis/src/hir_ty_lowering/errors.rs:
```diff
@@ -1118,6 +1118,14 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
         for (what, span) in types_and_spans {
             err.span_label(span, format!("not allowed on {what}"));
         }
+        if segments.clone().any(|segment| {
```
`segments is a `impl Iterator + Clone`, so cloning it is cheap and doesn't really allocate much 😅

If it was a `Vec`, then I wouldn't need to clone it, it'd be `.iter()` and nothing else.

lcnr reviewed (commented):
it's not that they aren't allowed on enum variants at all, they are only forbidden if they are already specified on the enum type:
```rust
type OptionU32 = Option<u32>;
fn main() {
    Some::<u32>; // ok
    Option::Some::<u32>; // ok
    Option::<u32>::Some::<u32>; // ERROR
    OptionU32::Some; // ok
    OptionU32::Some::<u32>; // ERROR
}
```

BoxyUwU commented:
@rustbot author

lcnr commented on compiler/rustc_hir_analysis/src/hir_ty_lowering/errors.rs:
```diff
@@ -1522,6 +1522,29 @@ fn generics_args_err_extend<'a>(
+                           path segments simultaneously; they are only valid in one place or the \
+                           other";
+                err.note(msg);
+                err.multipart_suggestions(
+                    "remove the generics arguments from one of the path segments",
+                    args.into_iter().map(|span| vec![(span, String::new())]),
+                    Applicability::MaybeIncorrect,
+                );
```
vibe: I feel like making this a suggestion is not worth it and that the 6 additional lines in the error message add more noise. I would personally prefer to only make this a help message

r=me with or without that change

estebank replied:
I had a few different alternatives on how to communicate this, because previously we didn't point at the previous type args:

```
error[E0109]: type arguments are not allowed on tuple variant `TSVariant`
  --> $DIR/enum-variant-generic-args.rs:54:29
   |
LL |     Enum::<()>::TSVariant::<()>(());
   |                 ---------   ^^ type argument not allowed
   |                 |
   |                 not allowed on tuple variant `TSVariant`
   |
   = note: generic arguments are not allowed on both an enum and its variant's path segments simultaneously; they are only valid in one place or the other
help: remove the generics arguments from one of the path segments
   |
LL -     Enum::<()>::TSVariant::<()>(());
LL +     Enum::TSVariant::<()>(());
   |
LL -     Enum::<()>::TSVariant::<()>(());
LL +     Enum::<()>::TSVariant(());
   |

error[E0109]: type arguments are not allowed on tuple variant `TSVariant`
  --> $DIR/enum-variant-generic-args.rs:54:29
   |
LL |     Enum::<()>::TSVariant::<()>(());
   |                 ---------   ^^ type argument not allowed
   |                 |
   |                 not allowed on tuple variant `TSVariant`
   |
   = note: generic arguments are not allowed on both an enum and its variant's path segments simultaneously; they are only valid in one place or the other
   |
LL |     Enum::<()>::TSVariant::<()>(());
   |           ^^^^              ^^

error[E0109]: type arguments are not allowed on tuple variant `TSVariant`
  --> $DIR/enum-variant-generic-args.rs:54:29
   |
LL |     Enum::<()>::TSVariant::<()>(());
   |           ----  ---------   ^^
   |           |     |           |
   |           |     |           type argument not allowed
   |           |     |           type argument on the variant
   |           |     not allowed on tuple variant `TSVariant`
   |           type argument on the enum
   |
   = note: generic arguments are not allowed on both an enum and its variant's path segments simultaneously; they are only valid in one place or the other
```

I wasn't happy with any of them, tbh. They are all verbose, at lesat the suggestion one has the chance of being understood by the widest amount of people. One other option is to emit a specific error:

```
error[E0109]: type arguments are not allowed on enum `Enum` and tuple variant `TSVariant` at the same time
  --> $DIR/enum-variant-generic-args.rs:54:29
   |
LL |     Enum::<()>::TSVariant::<()>(());
   |            ^^   ---------   ^^ type argument on the variant
   |            |     
   |            type argument on the enum
   |
   = note: generic arguments are not allowed on both an enum and its variant's path segments simultaneously; they are only valid in one place or the other
```

Any preference?

lcnr replied:
maybe also pointing to the generic args on the enum inline
```
error[E0109]: type arguments are not allowed on variant `SVariant`
  --> $DIR/enum-variant-generic-args.rs:72:28
   |
LL |     Enum::<()>::SVariant::<()> { v: () };
   |                 --------   ^^ type argument not allowed
   |                 |
   |                 not allowed on variant `SVariant`
   |
   = note: generic arguments are not allowed on both an enum and its variant's path segment simultaneously
   = help: consider removing the redundant generic arguments from the variant's path
```
I feel like we should be opinionated here and suggest keeping the args on the enum.

estebank replied:
What about the current output?

lcnr replied:
I personally prefer my suggestion from the last comment over the status quo. It's also still broken in the `AliasFixed` error message, isn't it?

I think stating both "not allowed on enum `Enum`" and "not allowed on tuple variant `TSVariant`" is confusing. I would chnage the first to "already specified on enum `Enum`" maybe?

fmease commented on compiler/rustc_hir_analysis/src/hir_ty_lowering/errors.rs:
```diff
@@ -1496,6 +1497,32 @@ fn generics_args_err_extend<'a>(
+                    _ => None,
+                })
+                .collect();
+            if args.len() > 1
+                && let Some(span) = args.into_iter().last()
+            {
```
```suggestion
            if let [.., arg] = *args {
```

BoxyUwU replied:
`..` can match 0 elements so `&[1]` will match a pattern `[.., 1]`, you would need `if let [_, .., span]` for this to do what you want i think. This seems somewhat weird to write whereas existing logic seems very straightforward to me

BoxyUwU commented on compiler/rustc_hir_analysis/src/hir_ty_lowering/errors.rs:
```diff
@@ -1047,23 +1047,27 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
             };
         });
 
+        let segments: Vec<_> = match err_extend {
+            GenericsArgsErrExtend::DefVariant(segments) => segments.iter().collect(),
+            _ => segments.collect(),
+        };
```
I think I agree with lcnr about how having extra spans and things in the error message for the enum arguments too is confusing. Can you just remove this so we keep the current set of segments, then I think this is good to merge

estebank replied:
done

BoxyUwU commented:
@rustbot author

estebank committed:
Remove some the spans pointing at the enum in the path and its generic args

```
error[E0109]: type arguments are not allowed on tuple variant `TSVariant`
  --> $DIR/enum-variant-generic-args.rs:54:29
   |
LL |     Enum::<()>::TSVariant::<()>(());
   |                 ---------   ^^ type argument not allowed
   |                 |
   |                 not allowed on tuple variant `TSVariant`
   |
   = note: generic arguments are not allowed on both an enum and its variant's path segments simultaneously; they are only valid in one place or the other
help: remove the generics arguments from one of the path segments
   |
LL -     Enum::<()>::TSVariant::<()>(());
LL +     Enum::<()>::TSVariant(());
   |
```

estebank commented:
@BoxyUwU changed

@rustbot review

BoxyUwU commented:
@bors r+ rollup
